### PR TITLE
OLMo Core to HF conversion refactor

### PR DESCRIPTION
**Pull request by 2015aroras**
Issue: Our conversion logic between OLMo Core and HF transformers formats is not very flexible. Every time we want to support a new architecture, we have to write new logic and test from scratch. The ideal is to have an easy way to specify how OLMo Core states maps to/from HF, and have our code use that mapping information to perform the conversion.

Fix: This PR refactors the logic around conversion to make it easier to convert between the formats. It enables us to specify mappings between states of the 2 formats that support placeholders for things like expert id and layer id (in `src/olmo_core/nn/hf/convert.py` using `StateMappingTemplate`s). The mappings can be many-to-many and allow us to specify some operations that need to be performed as part of conversion, like permuting dimensions or merging/unmerging layers. The operations currently supported are the minimum needed to support OLMo2 and the current MoE implementation.

### CODE REVIEW

File: src/olmo_core/nn/conversion/state_mapping.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    EXPERT = "[expert]"
+
+
+@dataclass
+class StateMappingTemplate:
+    """
+    The template for a mapping from state from one format to another format (e.g. OLMo Core to HF).
+    These mappings are 'templates' since they support keys and other metadata having placeholders
+    for information like the layer number or number of MoE experts. This class can be converted
+    to a `StateMapping` by providing the placeholder information information.
```
Comment (epwalsh) [inline src/olmo_core/nn/conversion/state_mapping.py]: minor typos:
```suggestion
    The template for a mapping state from one format to another format (e.g. OLMo Core to HF).
    These mappings are 'templates' since they support keys and other metadata having placeholders
    for information like the layer number or number of MoE experts. This class can be converted
    to a `StateMapping` by providing the placeholder information.
```

File: src/olmo_core/nn/conversion/state_mapping.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    The key or keys of the state(s) being mapping from.
+    """
+    dest_template_keys: str | Tuple[str, ...]
+    """
+    The key or keys of the state(s) being mapping to.
+    """
+
+    source_key_per_placeholder: TemplatePlaceholder | None = None
+    """
+    A placeholder in `source_template_keys` for which this mapping should map all valid placeholder
```
Comment (epwalsh) [inline src/olmo_core/nn/conversion/state_mapping.py]: Annoyingly ReStructured Text requires double backticks for code formatting.

```suggestion
    A placeholder in ``source_template_keys`` for which this mapping should map all valid placeholder
```

File: src/olmo_core/nn/conversion/state_converter.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+
+class StateConverter:
+    """
+    A class for converting state from one format to another format (e.g. OLMo Core to HF).
+    """
+
+    def __init__(self, mapping_templates: List[StateMappingTemplate]) -> None:
+        self.mapping_templates = mapping_templates
```
Comment (epwalsh) [inline src/olmo_core/nn/conversion/state_converter.py]: Could just make this a dataclass to avoid the `__init__()` ?

File: src/olmo_core/nn/hf/checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    barrier(group=process_group)
+
+    hf_model = AutoModelForCausalLM.from_pretrained(model_name_or_path)
+    log.info(f"Loaded hf model: {hf_model}")
+
+    converted_state_dict: Dict[str, torch.Tensor] = convert_state_from_hf(
+        hf_model.config, hf_model.state_dict(), model_id=model_id
+    )
+
+    for key, state in converted_state_dict.items():
```
Comment (epwalsh) [inline src/olmo_core/nn/hf/checkpoint.py]: Since there's potentially some distributed communication here you'll need to make sure the keys are traversed in the same order on each rank.
```suggestion
    for key in sorted(converted_state_dict.keys()):
        state = converted_state_dict[key]
```

File: src/examples/huggingface/convert_checkpoint_from_hf.py [hunk 5]
```diff
@@ -8,165 +8,372 @@
...
+        self.to_empty(device=device)
+
+        for module in self.modules():
+            if hasattr(module, "reset_parameters"):
+                module.to_empty(device=device)  # type: ignore
+
+        seed = self.init_seed
+        if pp_mesh is not None:
+            seed += pp_mesh.get_local_rank()
+        return torch.Generator(device).manual_seed(seed)
```
Comment (epwalsh) [inline src/examples/huggingface/convert_checkpoint_from_hf.py]: I don't think this is necessary. You could just call `model.to_empty(device=device)`.

File: src/examples/huggingface/convert_checkpoint_from_hf.py [hunk 6]
```diff
@@ -8,165 +8,372 @@
...
+
+    model.init_weights = types.MethodType(init_weights, model)
+
+    device = device or get_default_device()
+
+    train_module = TransformerTrainModuleConfig(
+        rank_microbatch_size=max_sequence_length,
+        max_sequence_length=max_sequence_length,
+        optim=AdamWConfig(),
+    ).build(model, device=device)
```
Comment (epwalsh) [inline src/examples/huggingface/convert_checkpoint_from_hf.py]: I know technically the train module decides the format of the checkpoint, but it seems overkill to have to instantiate a train module here. In this case I think it's fair to make assumptions about the format and just use the corresponding functions from `olmo_core.distributed.checkpoint` directly.

**Review by epwalsh — COMMENTED** [review]
Looks good overall, especially the `nn` API. I left a couple suggestions about how to simplify the example conversion scripts, and I have a couple questions:
- Does this cover removing padding from the embeddings?
- Can you add tests for the conversion scripts on a tiny model?

File: src/olmo_core/nn/conversion/state_converter.py (hunk 3 above)
Comment (2015aroras) [reply src/olmo_core/nn/conversion/state_converter.py]: Done https://github.com/allenai/OLMo-core/pull/226/commits/197bcca0ff42d125403722b663ea98b7d300f8aa

File: src/examples/huggingface/convert_checkpoint_from_hf.py (hunk 5 above)
Comment (2015aroras) [reply src/examples/huggingface/convert_checkpoint_from_hf.py]: This was necessary due to the use of `TransformerTrainModule`, but since that is gone (https://github.com/allenai/OLMo-core/pull/226/commits/8028909162812d74de16525f7132f0472f0cc4c0) I have fixed this too.

File: src/examples/huggingface/convert_checkpoint_from_hf.py (hunk 6 above)
Comment (2015aroras) [reply src/examples/huggingface/convert_checkpoint_from_hf.py]: Done https://github.com/allenai/OLMo-core/pull/226/commits/8028909162812d74de16525f7132f0472f0cc4c0

**Comment by 2015aroras** [comment]
> Does this cover removing padding from the embeddings?

WIP, since this was mentioned fairly recently

> Can you add tests for the conversion scripts on a tiny model?

👍

File: src/olmo_core/nn/conversion/state_mapping.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    EXPERT = "[expert]"
+
+
+@dataclass
+class StateMappingTemplate:
+    """
+    The template for a mapping state from one format to another format (e.g. OLMo Core to HF).
+    These mappings are 'templates' since they support keys and other metadata having placeholders
+    for information like the layer number or number of MoE experts. This class can be converted
+    to a `StateMapping` by providing the placeholder information.
```
Comment (epwalsh) [inline src/olmo_core/nn/conversion/state_mapping.py]:
```suggestion
    to a :class:`StateMapping` by providing the placeholder information.
```

File: src/olmo_core/nn/conversion/state_mapping.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+    The most standard mapping is a one-to-one state mapping, which corresponds to a single
+    string entry for both `source_template_keys` and `dest_template_keys`. The class also supports
```
Comment (epwalsh) [inline src/olmo_core/nn/conversion/state_mapping.py]: I still see a bunch of instances like this that should have double backticks. You can check how this renders by running `make docs`, though you may need to add a page for this submodule in `docs/source/nn/` (and link to it in `docs/source/nn/index.rst`).

```suggestion
    string entry for both ``source_template_keys`` and ``dest_template_keys``. The class also supports
```

**Review by epwalsh — COMMENTED** [review]
LGTM, just a couple more doc issues

File: src/olmo_core/nn/conversion/state_mapping.py (hunk 8 above)
Comment (2015aroras) [reply src/olmo_core/nn/conversion/state_mapping.py]: Done a mass update of docs
[b1b6e7e](https://github.com/allenai/OLMo-core/pull/226/commits/b1b6e7e4c40291fef1a03643a003207c2a21208e)

**Comment by 2015aroras** [comment]
@epwalsh Gentle ping.

File: pyproject.toml [hunk 9]
```diff
@@ -20,6 +20,7 @@ dependencies = [
     "packaging",
     "rich",
     "omegaconf",
+    "transformers",
```
Comment (epwalsh) [inline pyproject.toml]: Can we make this as an optional dependency? I think the HF stuff is pretty self-contained, right?

Comment (2015aroras) [reply pyproject.toml]: https://github.com/allenai/OLMo-core/pull/226/commits/ddca1b0892d006d332d3c8308ccf933d5dd90eb0
